cmd/eksctl-anywhere/cmd/downloadartifacts.go

```diff
@@ -0,0 +1,206 @@
+		if err := downloadArtifacts(cmd.Context(), downloadArtifactsopts); err != nil {
+			return err
+		}
+		return nil
+	},
+}
+
+func downloadArtifacts(context context.Context, opts *downloadArtifactsOptions) error {
+	cliVersion := version.Get()
+	clusterSpec, err := cluster.NewSpec(opts.fileName, cliVersion)
+	if err != nil {
+		return err
+	}
+
+	release, err := clusterSpec.GetRelease(cliVersion)
+	if err != nil {
+		return err
+	}
+	bundlesManifestUrl := release.BundleManifestUrl
+
+	specManifests := []string{
+		bundlesManifestUrl,
+		clusterSpec.GetReleaseManifestUrl(),
+	}
+	for _, manifestURI := range specManifests {
+		if opts.dryRun {
+			logger.Info(fmt.Sprintf("Found artifact: %s\n", manifestURI))
+			continue
+		}
+		if err = downloadArtifact("", opts.downloadDir, manifestURI, clusterSpec); err != nil {
+			return fmt.Errorf("error downloading artifact: %v", err)
+		}
+	}
+
+	bundle := clusterSpec.VersionsBundle
+	for component, manifestList := range bundle.Manifests() {
+		for _, manifest := range manifestList {
+			if opts.dryRun {
+				logger.Info(fmt.Sprintf("Found artifact: %s\n", manifest.URI))
+				continue
+			}
+			if err = downloadArtifact(component, opts.downloadDir, manifest.URI, clusterSpec); err != nil {
+				return fmt.Errorf("error downloading artifact for component %s: %v", component, err)
+			}
+		}
+	}
+
+	if !opts.dryRun {
+		if err = createTarball(opts.downloadDir); err != nil {
+			return err
+		}
+
+		if !opts.retainDir {
+			if err = os.RemoveAll(opts.downloadDir); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func preRunDownloadArtifactsCmd(cmd *cobra.Command, args []string) error {
+	cmd.Flags().VisitAll(func(flag *pflag.Flag) {
+		if err := viper.BindPFlag(flag.Name, flag); err != nil {
+			log.Fatalf("Error initializing flags: %v", err)
+		}
+	})
+	return nil
+}
+
+func downloadArtifact(component, downloadDir, artifactUri string, s *cluster.Spec) error {
+	logger.V(3).Info(fmt.Sprintf("Downloading artifact: %s", artifactUri))
+
+	fileName := filepath.Base(artifactUri)
+
+	var filePath string
+	if component != "" {
+		filePath = filepath.Join(downloadDir, component, fileName)
+	} else {
+		filePath = filepath.Join(downloadDir, fileName)
+	}
+	if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
+		return err
+	}
+
+	logger.V(3).Info(fmt.Sprintf("Creating local artifact file: %s", filePath))
+
+	contents, err := s.ReadFile(artifactUri)
+	if err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(filePath, contents, 0o644); err != nil {
+		return err
+	}
+
+	logger.V(3).Info(fmt.Sprintf("Successfully downloaded artifact %s to %s", artifactUri, filePath))
+
+	return nil
+}
+
+func createTarball(downloadDir string) error {
+	var buf bytes.Buffer
+	tarFileName := fmt.Sprintf("%s.tar.gz", downloadDir)
+	tarFile, err := os.Create(tarFileName)
+	if err != nil {
+		return err
+	}
+	defer tarFile.Close()
+
+	gzipWriter := gzip.NewWriter(&buf)
```

Comment:
should we also defer close gzipWriter and tarWriter since we check for error and return on line 189

Reply:
Yup, good call. Thanks!